README.md
# AntSword [![release](https://img.shields.io/badge/release-v2.0.6-blue.svg?style=flat-square)][url-release]

> AntSword in your hands, no worries in your mind！

**AntSword is an open source, cross-platform website administration tool, being designed to meet the needs of penetration testers together with security researchers with permissions and/or authorizations as well as webmasters.**

**Anyone shall not use it for illegal purposes and profitability. Besides that, publishing unauthorized modified version is also prohibited, or otherwise bear legal responsibilities.**

This software, of which the development thought is modularization, is intended to provide easy-to-understand codes and modification guidelines for users of different levels. Therefore, any contribution making by everyone to this project is encouraged, whether large or small. By doing so, this tool can be more convenient and consequently become your most powerful kit！

[中文说明][url-doczh] / [Document][url-document] / [Changelog][url-changelog]

## Development stack

- [Electron][url-electron]
- [ES6][url-es6]
- [dhtmlx][url-dhtmlx]
- [Nodejs][url-nodejs]
* And other libraries called in the project.

## Screenshots

![][url-mainui]

<details>

<summary>More Screenshots</summary>

![][url-filemanager]
![][url-terminal]
![][url-database]
![][url-pluginstore]

</details>

## Quick Start

See document: [Quick Start][url-quickstart]

## Contribute

See document: [Make contributions to AntSword][url-contribute]

## Thanks

> Thanks to anyone who made any contributions.

## LICENSE

[LICENSE](./LICENSE)

[url-doczh]: README_CN.md
[url-changelog]: CHANGELOG.md
[url-document]: http://doc.u0u.us/en/
[url-release]: https://github.com/AntSwordProject/AntSword/releases/
[url-electron]: http://electron.atom.io/
[url-es6]: http://es6.ruanyifeng.com/
[url-dhtmlx]: http://dhtmlx.com/
[url-nodejs]: https://nodejs.org/
[url-homepage]: http://uyu.us
[url-release]: https://github.com/AntSwordProject/AntSword/releases
[url-quickstart]: https://doc.u0u.us/en/getting_started/index.html
[url-contribute]: https://doc.u0u.us/en/contribute_docs.html
[url-mainui]: http://as.xuanbo.cc/doc/getting_started/get_antsword_2.jpg
[url-filemanager]: http://as.xuanbo.cc/doc/file_manager/main_page_1.png
[url-terminal]: http://as.xuanbo.cc/doc/terminal/main_page_1.png
[url-database]: http://as.xuanbo.cc/doc/database/main_page_1.png
[url-pluginstore]: http://as.xuanbo.cc/doc/plugin_store/main_page_2.jpg


CHANGELOG.md
# 更新日志
> 有空会补补BUG、添添新功能。    
> 同时也欢迎大家的参与！感谢各位朋友的支持！ .TAT.

## `v(2.0.7)`

### 核心

* 执行命令新增 `antsystem` 函数, 具体见: https://github.com/AntSwordProject/ant_php_extension
* Fix 编码器保存后不生效的问题 #135 (thx @K4ngx)
* Fix core download 参数遗漏问题 #142 (thx @RoyTse)
* Fix #141 (thx @RoyTse)

### 虚拟终端

* 新增自定义命令 `aslistcmd`, 列出可使用的命令解释器 (#123)

![aslistcmd.png](https://i.loli.net/2019/04/05/5ca74815a01a4.png)
![aslistcmd_win.png](https://i.loli.net/2019/04/05/5ca74819b276e.png)

* 新增自定义命令 `aspowershell [on|off]`, 开启/关闭 PowerShell 模式

> 如果`ascmd`命令指定的PowerShell解释器文件名中包函`powershell`关键字, 会自动启用 PowerShell 模式, 如下图:

![aspowershell_default](https://i.loli.net/2019/04/05/5ca753673efe7.png)

> 如果指定的 PowerShell 解释器文件名中不包含 `powershell` 关键字, 则需要手动使用该命令，启用 PowerShell 模式。如果关闭了 PowerShell 模式,则会执行出错，如下图:

![aspowershell_switch](https://i.loli.net/2019/04/05/5ca75368d85fa.png)

### 其它

* 修复默认设置保存时导致 bookmarks 清空的问题
* PHP Custom Shell 新增 listcmd 功能
* PHP Custom Shell 新增数据库支持函数检查接口
* 新增「繁體中文(香港)」和「繁體中文(台灣)」两种语言

## 2019/03/20 `v(2.0.6)`

### 后端模块

* 分块传输自动根据黑名单字符(eg: eval, assert, execute, response 等)进行随机切割(thx @phith0n)

### 数据库管理

* 新增「测试连接」功能
* 新增「检测」功能, 检测支持的数据库函数(目前仅 PHP,ASP,ASPX 有效, ASP(X)仅检测使用到的组件是否存在)
* 新增 php `sqlsrv` 连接方式, php5.3之后 mssql 默认不存在,可使用该类型连接 sqlserver >= 2008

> 如果直连shell本地sqlserver, host 部分填 localhost 或者 (local)
>
> 如果连接外部,使用 ip,port

* 新增 php `oracle_oci8` 连接方式, 用于连接 oracle 8i,9i,10g,11g,12c

> host 部分填写: localhost/orcl 或者 localhost:1521/orcl
>
> 参考: http://php.net/manual/zh/function.oci-connect.php  connection_string 部分

* 新增 PHP PostgreSQL 类型数据库操作
* 新增 PHP PostgreSQL_PDO 类型数据库操作 #133 (thx @B0y1n4o4)
* 优化 asp(x) Oracle 类型数据库操作
* 优化 asp(x) SQLServer 类型数据库操作
* 优化SQLServer类型数据库默认查询语句
* php数据管理解析数据时自动猜解编码

### 文件管理

* 新增「在此处打开终端」功能, 打开终端后快速跳到当前目录下
* 新增「全局书签」功能, 可在「系统设置-默认设置」单击鼠标右键添加

![](https://i.loli.net/2019/03/13/5c891b279c26a.png)

![](https://i.loli.net/2019/03/13/5c891b2cc1c30.png)

### 数据管理

* shell 配置页面提示不推荐使用 default、random 编码器, 明文传输 Payload 容易受到转义等影响，未来版本将会考虑移除
* 新增「创建副本」菜单, 复制所选择的 Shell 并在相同分类下创建一个副本
* 新增「搜索数据」功能, 搜索本地数据，范围为当前分类下的 Shell

可选搜索字段： URL(URL地址), Password(密码), Remark(备注), All(在以上几个字段中出现)

> 如果想搜索其它类型的字段, 可直接在类型框中输入想要搜索的字段
>
> 例如搜索 IP地址, 可在搜索字段选择框中输入: ip
>
> 具体字段类型名请直接查阅 antData/db.ant

唤醒快捷键 Ctrl+Shift+F 或者 Command + Shift + F (OSX)

退出:

1) 点击搜索框之外的任何区域

2) 按下 `ESC` 键

3) 再次按下唤醒快捷键

> 在使用快捷键时,如果当前活动 tab 不是数据管理，则会自动跳回数据管理

### 其它

* 新增 「JSP Custom Shell For Oracle」
* 新增 Decodes 自动猜解编码,在中文少量的情况下,成功率会降低
* 系统托盘新增「重启应用」菜单
* 虚拟终端打开后提示本地命令菜单 ashelp

### BugFix

* 修复 asp(x) sqlserver 获取列名,执行自定义SQL语句的异常
* 修复 php mssql 获取列名,执行自定义SQL语句异常

## 2019/03/04 `v(2.0.5)`

### 后端模块

* 新增 chunked 流式传输

![chunk-2.png](https://i.loli.net/2019/01/28/5c4e8869c83ae.png)

### Shell 管理

* 新增「测试连接」功能
* 其它设置增加「开启分块传输」开关，设置「最小分块」和「最大分块」。每次分块大小为[n, m]区间。

![chunk-1.png](https://i.loli.net/2019/01/28/5c4e8868f178a.png)

### 文件管理

* 编辑文件标签打开时显示正在编辑的文件的「IP地址」和「绝对路径」 (#129)
* 编辑文件「保存」按钮移动到右侧

### 插件市场

* 支持使用代理访问插件市场,当配置了代理之后，默认开启

### 浏览网站

* 新增代理开关, 在配置了代理之后，可通过该开关控制是否使用代理浏览目标网站

### Other

* 调整 aproxy uri 初始化 URL 格式

## 2019/01/21 `v(2.0.4)`

### 后端模块

 * 优化返回包解码流程,增加自动猜解编码功能,该功能在中文较少时容易出错,出错了使用用户设置的字符编码

### 文件管理

 * 编辑文件窗口将「编码」按钮替换为「用此编码打开」, 若打开文件之后乱码，不妨试试切换编码
 * 新增「复制文件名」和「复制文件路径」功能，目标复制文件名或路径到剪贴板
 * 编辑文件菜单可选择「标签打开」和「窗口打开」

### 虚拟终端

 * 新增两条本地命令
  * ashelp 列出所有本地命令
  * ascmd  指定某个文件执行后续的命令. eg: ascmd /bin/bash 这样后续输入的命令将会使用 /bin/bash 来执行。注意仅对当前有效, 如果需要持久生效, 需要在shell配置中的「其它配置」设置。

### 系统设置

 * 新增「默认设置」, 可在此配置文件管理默认编辑文件打开方式

### Other

 * 新增 Python2 Custom CGI shell 示例

### BugFix

 * 修复Linux中编码设置中找到不到路径的问题 (#117) thx @H1d3r
 * 修复 Windows shell 盘符小写引起文件管理目录树错乱的 Bug
 * 修复 ASP 文件创建失败和删除失败时提示

### 插件

#### ASRedis

* 新增「虚拟命令行」功能

 ![](https://i.loli.net/2019/01/16/5c3f2134dcd8f.png)

## 2018/12/25 `(v2.0.3)`

### 模块增强

#### Shell 管理

 * 添加 Shell 时 URL 默认前缀 `http://`
 * 添加 Shell 时根据文件后缀选择 Shell 类型并赋予默认编码(asp 默认 GBK, 其它默认 UTF8) #109
 * 其它配置新增 `Multipart 发包` 功能

#### 后端模块

 * 数据存储新增插件配置存储管理功能 (`shell-addPluginDataConf`, `shell-editPluginDataConf`, `shell-delPluginDataConf`, `shell-getPluginDataConf`)
 * 后台发包方式支持 `Multipart`, 可在「编辑Shell配置」-「其它配置」里选择是否开启此功能，默认关闭。(thx @phith0n)

### Bug Fix

 * 修复数据库编码无法保存的 Bug (#110 thx @Twi1ight)
 * 修复 PHP Mysql(i) 数据管理模版代码中编码设置部分的错误 (#110 thx @Twi1ight)

### Other

 * 自动检查更新每24小时触发一次(GitHub 访问频率限制)
 * 插件市场默认窗口大小调整
 * Loading 界面增加了圣诞节彩蛋, 偶尔跟风过个节 🎄 Merry Christmas 🎄

## 2018/12/05 `(v2.0.2)`

### 模块增强

#### Shell 管理

 * 可在 Shell 编辑数据窗口下的`其它设置`栏目下自定义上传文件分片大小 (默认为 500KB, 太大会导致 HTTP 413 错误)

#### 文件管理

 * PHP Shell 下可直接修改文件权限, 显示为4位 8进制数, 如 `0644`

#### 数据管理
 
 * 优化了查询结果显示，默认所有列宽为 100
 * 可将查询结果导出为 CSV 文件
 * PHP Shell MySQL 数据库可视化增强，支持`新建数据库`,`删除数据库`,`新建表`,`修改表名`,`删除表`,`编辑列名`,`删除列`

#### 虚拟终端

 * 虚拟终端界面下使用 `Ctrl` + `=`(和`+`在一起的那个键) 可放大, `Ctrl` + `-` 可缩小

#### 浏览网站

 * 新增了地址栏, 面对需要先进入登录页面的 Shell, 可先在此处访问 login 页面，然后保存 Cookie 到 Shell 配置。 默认为 Shell 的 URL
 * 调整了工具栏按钮的排列
 * 关闭了默认自动打开 URL,需要手动点击「浏览」按钮

### Bug Fix

 * 修正 windows 客户端下用户编码器路径解析错误的问题

### Other

 * 数据分割符随机化，再也不是之前固定的 `->|` 和 `|<-` 了
 * 支持返回状态为 404, 500, 403 等非 200 的 Shell (#103 thx @Curz0n)，一个简单的例子如下：
  
  ```
    <?php http_response_code(404);@eval($_POST['ant']);?>
  ```

 * JSP Shell 基础信息调整, 现在默认的目录为 shell 编译后的 class 文件所在目录
 * 关于页面新增 [Discord 在线交流地址](https://discord.gg/Uzh5nUf)

## 2018/09/12 `(v2.0.1)`

### 插件

* 新增 [「LiveScan」](https://github.com/virink/LiveScan) 插件 (Author: @virink)

 > Webshell存活弹出插件

 1. 通过请求 Webshell 并判断返回数据是否一致判断 Webshell 是否存活
 2. 仅对 PHP,ASP,ASPX 有效
 3. 一键将失联的 Webshell 移动到 `[.Trash]` 分类
 4. 一键清空 `[.Trash]` 分类

### Bug Fix

* 修正下载更新时 chunk 导致进度条显示问题
* 修正 Windows 客户端用户自定义编码器加载问题
* 修正 C 盘无权限目录显示的 Bug(#51)
* 修正 Windows 客户端下上传文件路径截取问题(#97 thx @jjf012)
* 修正 Linux 客户端载入源码路径错误的问题 (cbb67dd0)
* 修正 Windows 服务器修改文件时间失败问题(#98 thx @jjf012)
* 修正 ASP Access 数据库显示数据库失败问题

### Other

* PHP虚拟终端执行命令函数增加了 `shell_exec`, `passthru`, `exec`, `popen`
* ASP wget 功能模版 `Microsoft.XMLHTTP` 替换为 `MSXML2.ServerXmlHttp` 支持 https (#98 thx @jjf012)
* 添加「检查更新」菜单
* 提示通过 git 控制源码用户通过 git 命令增量更新(http下载的源码暂时不支持增量更新)

## 2018/08/25 `(v2.0.0)`


modules/database.js
/**
 * Shell数据库管理模块
 * 更新：2016/06/28
 */

'use strict';

const fs = require('fs'),
  dns = require('dns'),
  path = require('path'),
  CONF = require('./config'),
  Datastore = require('nedb'),
  qqwry = require("geoips").info();

var logger;

class Database {

  /**
   * 初始化数据库
   * @param  {electron} electron electron对象
   * @return {[type]}          [description]
   */
  constructor(electron) {
    logger = new electron.Logger('Database');
    this.cursor = new Datastore({
      filename: CONF.dataPath,
      autoload: true
    });
    // 监听事件
    electron.ipcMain
      .on('shell-add', this.addShell.bind(this))
      .on('shell-del', this.delShell.bind(this))
      .on('shell-edit', this.editShell.bind(this))
      .on('shell-move', this.moveShell.bind(this))
      .on('shell-find', this.findShell.bind(this))
      .on('shell-clear', this.clearShell.bind(this))
      .on('shell-findOne', this.findOneShell.bind(this))
      .on('shell-addDataConf', this.addDataConf.bind(this))
      .on('shell-editDataConf', this.editDataConf.bind(this))
      .on('shell-delDataConf', this.delDataConf.bind(this))
      .on('shell-getDataConf', this.getDataConf.bind(this))
      .on('shell-renameCategory', this.renameShellCategory.bind(this))
      .on('shell-updateHttpConf', this.updateHttpConf.bind(this))
      .on('shell-addPluginDataConf', this.addPluginDataConf.bind(this))
      .on('shell-editPluginDataConf', this.editPluginDataConf.bind(this))
      .on('shell-delPluginDataConf', this.delPluginDataConf.bind(this))
      .on('shell-getPluginDataConf', this.getPluginDataConf.bind(this));
  }

  convertOptstoNedbQuery(opts={}) {
    var self = this;
    if(opts instanceof Array) {
      for (let i = 0; i < opts.length; i++) {
        opts[i] = self.convertOptstoNedbQuery(opts[i]);
      }
    }else if(opts instanceof Object) {
      Object.keys(opts).map((f) => {
        if(opts[f] instanceof Object) {
          opts[f] = self.convertOptstoNedbQuery(opts[f]);
        }
        if(f == "$regex") {
          if(opts[f].charAt(0) == '*') {
            opts[f] = opts[f].substring(1);
          }
          opts[f] = new RegExp(opts[f], 'i');
        }
      });
    }
    return opts;
  }
  /**
   * 查询shell数据
   * @param  {Object} event ipcMain对象
   * @param  {Object} opts  查询配置
   * @return {[type]}       [description]
   */
  findShell(event, opts = {}) {
    opts = this.convertOptstoNedbQuery(opts);
    logger.debug('findShell', opts);
    this.cursor
      .find(opts)
      .sort({
        utime: -1
      })
      .exec((err, ret) => {
        event.returnValue = ret || [];
      });
  }

  /**
   * 查询单一shell数据
   * @param  {Object} event ipcMain对象
   * @param  {String} opts  shell id
   * @return {[type]}       [description]
   */
  findOneShell(event, opts) {
    logger.debug('findOneShell', opts);
    this.cursor.findOne({
      _id: opts
    }, (err, ret) => {
      event.returnValue = err || ret;
    });
  }


  /**
   * 根据URL解析出IP&&地理位置
   * @param  {String} url URL地址
   * @return {Promise}     ip, addr
   */
  _url2ip(url) {
    return new Promise((res, rej) => {
      // 解析domain
      const urlArr = url.match(/(\w+):\/\/([\w\.\-]+)[:]?([\d]*)([\s\S]*)/i);
      // 无效url
      if (!urlArr || urlArr.length < 3) {
        return rej('Unable to resolve domain name from URL');
      }
      // 获取IP
      const domain = urlArr[2];
      dns.lookup(domain, (err, ip) => {
        if (err) {
          return rej(err.toString());
        }
        // 获取地理位置
        const _addr = qqwry.searchIP(ip);
        return res({
          ip: ip,
          addr: `${_addr.Country} ${_addr.Area}`
        });
      })
    })
  }

  /**
   * 添加shell数据
   * @param {Object} event ipcMain对象
   * @param {Object} opts  数据（url,category,pwd,type,encode,encoder
   */
  addShell(event, opts) {
    logger.info('addShell', opts);
    if(opts.base['url'].match(CONF.urlblacklist)){
      event.returnValue = "Blacklist URL"
      return
    }
    this._url2ip(opts.base['url'])
      .then((ret) => {
        this.cursor.insert({
          category: opts.base['category'] || 'default',
          url: opts.base['url'],
          pwd: opts.base['pwd'],
          note: opts.base['note'],
          type: opts.base['type'],
          ip: ret['ip'],
          addr: ret['addr'],
          encode: opts.base['encode'],
          encoder: opts.base['encoder'],
          httpConf: opts.http,
          otherConf: opts.other,
          ctime: +new Date,
          utime: +new Date
        }, (_err, _ret) => {
          event.returnValue = _err || _ret;
        });
      })
      .catch((_err) => {
        event.returnValue = _err;
      })
  }

  /**
   * 编辑shell数据
   * @param  {Object} event ipcMain对象
   * @param  {Object} opts  数据（old,new
   * @return {[type]}       [description]
   */
  editShell(event, opts) {
    logger.warn('editShell', opts);

    const _new = opts.new;
    const _old = opts.old;
    if(_new.base['url'].match(CONF.urlblacklist)){
      event.returnValue = "Blacklist URL"
      return
    }
    this._url2ip(_new.base['url'])
      .then((ret) => {
        this.cursor.update({
          _id: _old['_id']
        }, {
          $set: {
            ip: ret['ip'],
            addr: ret['addr'],
            url: _new.base['url'],
            pwd: _new.base['pwd'],
            note: _new.base['note'],
            type: _new.base['type'],
            encode: _new.base['encode'],
            encoder: _new.base['encoder'],
            httpConf: _new.http,
            otherConf: _new.other,
            utime: +new Date
          }
        }, (_err, _ret) => {
          event.returnValue = _err || _ret;
        })
      })
      .catch((_err) => {
        event.returnValue = _err;
      });
  }

  /**
   * 更新httpConf配置信息（包含body&&headers
   * @param  {[type]} event [description]
   * @param  {[type]} opt  = {} [description]
   * @return {[type]}       [description]
   */
  updateHttpConf(event, opt = {}) {
    logger.warn('updateHttpConf', opt);

    this.cursor.update({
      _id: opt._id
    }, {
      $set: {
        httpConf: opt.conf,
        utime: +new Date
      }
    }, (_err, _ret) => {
      event.returnValue = _err || _ret;
    });
  }

  /**
   * 删除shell数据
   * @param  {Object} event ipcMain对象
   * @param  {Array}  opts  要删除的shell-id列表
   * @return {[type]}       [description]
   */
  delShell(event, opts) {
    logger.warn('delShell', opts);
    this.cursor.remove({
      _id: {
        $in: opts
      }
    }, {
      multi: true
    }, (err, num) => {
      event.returnValue = err || num;
    })
  }

  /**
   * 删除分类shell数据
   * @param  {Object} event ipcMain对象
   * @param  {String} opts  shell分类名
   * @return {[type]}       [description]
   */
  clearShell(event, opts) {
    logger.fatal('clearShell', opts);
    this.cursor.remove({
      category: opts
    }, {
      multi: true
    }, (err, num) => {
      event.returnValue = err || num;
    })
  }

  /**
   * 重命名shell分类
   * @param  {Object} event ipcMain对象
   * @param  {Object} opts  配置（oldName,newName
   * @return {[type]}       [description]
   */
  renameShellCategory(event, opts) {
    logger.warn('renameShellCategory', opts);
    this.cursor.update({
      category: opts['oldName']
    }, {
      $set: {
        category: opts['newName']
      }
    }, {
      multi: true
    }, (err, num) => {
      event.returnValue = err || num;
    })
  }

  /**
   * 移动shell数据分类
   * @param  {Object} event ipcMain对象
   * @param  {Object} opts  配置（ids,category
   * @return {[type]}       [description]
   */
  moveShell(event, opts) {
    logger.info('moveShell', opts);
    this.cursor.update({
      _id: {
        $in: opts['ids'] || []
      }
    }, {
      $set: {
        category: opts['category'] || 'default',
        utime: +new Date
      }
    }, {
      multi: true
    }, (err, num) => {
      event.returnValue = err || num;
    })
  }

  /**
   * 添加数据库配置
   * @param {Object} event ipcMain对象
   * @param {Object} opts  配置（_id,data
   */
  addDataConf(event, opts) {
    logger.info('addDataConf', opts);
    // 1. 获取原配置列表
    this.cursor.findOne({
      _id: opts['_id']
    }, (err, ret) => {
      let confs = ret['database'] || {};
      // 随机Id（顺序增长
      const random_id = parseInt(+new Date + Math.random() * 1000).toString(16);
      // 添加到配置
      confs[random_id] = opts['data'];
      // 更新数据库
      this.cursor.update({
        _id: opts['_id']
      }, {
        $set: {
          database: confs,
          utime: +new Date
        }
      }, (_err, _ret) => {
        event.returnValue = random_id;
      });
    });
  }

  /**
   * 修改数据库配置
   * @param {Object} event ipcMain对象
   * @param {Object} opts  配置（_id,id,data
   */
  editDataConf(event, opts) {
    logger.info('editDataConf', opts);
    // 1. 获取原配置列表
    this.cursor.findOne({
      _id: opts['_id']
    }, (err, ret) => {
      let confs = ret['database'] || {};
      // 添加到配置
      confs[opts['id']] = opts['data'];
      // 更新数据库
      this.cursor.update({
        _id: opts['_id']
      }, {
        $set: {
          database: confs,
          utime: +new Date
        }
      }, (_err, _ret) => {
        event.returnValue = opts['id'];
      });
    });
  }

  /**
   * 删除数据库配置
   * @param  {Object} event ipcMain对象
   * @param  {Object} opts  配置（_id,id
   * @return {[type]}       [description]
   */
  delDataConf(event, opts) {
    logger.info('delDataConf', opts);
    // 1. 获取原配置
    this.cursor.findOne({
      _id: opts['_id']
    }, (err, ret) => {
      let confs = ret['database'] || {};
      // 2. 删除配置
      delete confs[opts['id']];
      // 3. 更新数据库
      this.cursor.update({
        _id: opts['_id']
      }, {
        $set: {
          database: confs,
          utime: +new Date
        }
      }, (_err, _ret) => {
        event.returnValue = _err || _ret;
      });
    })
  }

  /**
   * 获取单个数据库配置
   * @param  {Object} event ipcMain对象
   * @param  {Object} opts  配置（_id,id
   * @return {[type]}       [description]
   */
  getDataConf(event, opts) {
    logger.info('getDatConf', opts);
    this.cursor.findOne({
      _id: opts['_id']
    }, (err, ret) => {
      const confs = ret['database'] || {};
      event.returnValue = err || confs[opts['id']];
    });
  }

  /**
   * 添加插件数据配置
   * @param {Object} event ipcMain对象
   * @param  {string} plugin 插件注册的名称
   * @param {Object} opts  配置（_id,data
   */
  addPluginDataConf(event, plugin, opts) {
    logger.info('addPluginDataConf', plugin, opts);
    // 1. 获取原配置列表
    this.cursor.findOne({
      _id: opts['_id']
    }, (err, ret) => {
      ret['plugins'] = ret['plugins'] || {};
      let confs = ret['plugins'][plugin] || {};
      // 随机Id（顺序增长
      const random_id = parseInt(+new Date + Math.random() * 1000).toString(16);
      // 添加到配置
      confs[random_id] = opts['data'];
      let setdata = {
        utime: +new Date,
      }
      setdata[`plugins.${plugin}`] = confs;

      // 更新
      this.cursor.update({
        _id: opts['_id']
      }, {
        $set: setdata
      }, (_err, _ret) => {
        event.returnValue = random_id;
      });
    });
  }
  /**
   * 修改插件数据配置
   * @param {Object} event ipcMain对象
   * @param  {string} plugin 插件注册的名称
   * @param {Object} opts  配置（_id,id,data
   */
  editPluginDataConf(event, plugin, opts) {
    logger.info('editPluginDataConf', plugin, opts);
    // 1. 获取原配置列表
    this.cursor.findOne({
      _id: opts['_id']
    }, (err, ret) => {
      ret['plugins'] = ret['plugins'] || {};
      let confs = ret['plugins'][plugin] || {};
      // 添加到配置
      confs[opts['id']] = opts['data'];
      let setdata = {
        utime: +new Date,
      }
      setdata[`plugins.${plugin}`] = confs;
      // 更新数据库
      this.cursor.update({
        _id: opts['_id']
      }, {
        $set: setdata
      }, (_err, _ret) => {
        event.returnValue = opts['id'];
      });
    });
  }

  /**
   * 删除插件数据配置
   * @param  {Object} event ipcMain对象
   * @param  {string} plugin 插件注册的名称
   * @param  {Object} opts  配置（_id,id
   * @return {[type]}       [description]
   */
  delPluginDataConf(event, plugin, opts) {
    logger.info('delPluginDataConf', plugin, opts);
    // 1. 获取原配置
    this.cursor.findOne({
      _id: opts['_id']
    }, (err, ret) => {
      ret['plugins'] = ret['plugins'] || {};
      let confs = ret['plugins'][plugin] || {};
      // 2. 删除配置
      delete confs[opts['id']];
      let setdata = {
        utime: +new Date,
      }
      setdata[`plugins.${plugin}`] = confs;
      // 3. 更新数据库
      this.cursor.update({
        _id: opts['_id']
      }, {
        $set: setdata
      }, (_err, _ret) => {
        event.returnValue = _err || _ret;
      });
    })
  }

  /**
   * 获取单个插件数据配置
   * @param  {Object} event ipcMain对象
   * @param  {string} plugin 插件注册的名称
   * @param  {Object} opts  配置（_id,id
   * @return {[type]}       [description]
   */
  getPluginDataConf(event, plugin, opts) {
    logger.info('getPluginDatConf', plugin, opts);
    this.cursor.findOne({
      _id: opts['_id']
    }, (err, ret) => {
      ret['plugins'] = ret['plugins'] || {};
      const confs = ret['plugins'][plugin] || {};
      event.returnValue = err || confs[opts['id']];
    });
  }
}

module.exports = Database;
